File: TestPlugin/src/main/java/xyz/wingio/plugins/testplugin/PermChip.java
package xyz.wingio.plugins.testplugin;

import android.content.Context;
import android.view.*;
import android.widget.*;
import android.graphics.drawable.*;
import android.graphics.*;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;

import com.discord.utilities.color.ColorCompat;
import com.aliucord.Utils;
import com.facebook.drawee.view.SimpleDraweeView;
import com.google.android.material.card.MaterialCardView;
import com.lytefast.flexinput.R;

public class PermChip extends MaterialCardView {
    public TextView text;

    public PermChip(Context ctx, String value, int roleColor) {
        super(ctx);
        var clr = Color.parseColor("#" + String.format("%06x", roleColor));
        int p = Utils.dpToPx(8);
        LayoutParams params = new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
        setLayoutParams(params);
        setRadius(Utils.dpToPx(4));
        setCardBackgroundColor(ColorCompat.getThemedColor(ctx, R.b.colorBackgroundTertiary));

        LinearLayout root = new LinearLayout(ctx);
        root.setOrientation(LinearLayout.HORIZONTAL);
        root.setVerticalGravity(Gravity.CENTER_VERTICAL);
        root.setLayoutParams(new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT));
        root.setPadding(Utils.dpToPx(8), Utils.dpToPx(6), Utils.dpToPx(8), Utils.dpToPx(6));

        text = new TextView(ctx, null, 0, R.h.UiKit_TextAppearance_Semibold);
        text.setTextSize(12);
        text.setMaxLines(1);
        text.setText(value);

        ImageView dot = new ImageView(ctx);
        Drawable dotImg = ContextCompat.getDrawable(ctx, R.d.drawable_circle_white_12dp);
        if(roleColor != 0) dotImg.setTint(clr);
        dot.setImageDrawable(dotImg);
        LayoutParams dotParams = new LayoutParams(Utils.dpToPx(12), Utils.dpToPx(12));
        dotParams.setMarginEnd(Utils.dpToPx(8));
        dot.setLayoutParams(dotParams);
        
        root.addView(dot);
        root.addView(text);
        addView(root);
    }
}

File: ShowPermsV2/src/main/kotlin/xyz/wingio/plugins/showperms/view/PermChip.kt
package xyz.wingio.plugins.showperms.view

import android.content.Context
import android.graphics.Color
import android.view.Gravity
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.aliucord.utils.DimenUtils.dp
import com.discord.utilities.color.ColorCompat
import com.google.android.material.card.MaterialCardView
import com.lytefast.flexinput.R
import xyz.wingio.plugins.showperms.util.Settings

/**
 * A chip that is designed to emulate the appearance of the role chip
 */
class PermChip(
    private val ctx: Context
): MaterialCardView(ctx) {

    private val label: TextView
    private val roleDot: ImageView

    init {
        layoutParams = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
        radius = 4.dp.toFloat()
        setCardBackgroundColor(ColorCompat.getThemedColor(ctx, R.b.colorBackgroundTertiary))

        val root = LinearLayout(ctx).apply {
            setVerticalGravity(Gravity.CENTER_VERTICAL)
            setPadding(8.dp, 6.dp, 8.dp, 6.dp)
            orientation = LinearLayout.HORIZONTAL
            layoutParams = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
        }

        label = TextView(ctx, null, 0, R.i.UiKit_TextAppearance_Semibold).apply {
            textSize = 12f
            maxLines = 1
        }

        roleDot = ImageView(ctx).apply {
            layoutParams = LinearLayout.LayoutParams(12.dp, 12.dp).apply {
                marginEnd = 8.dp
            }
        }
        setDotColor(0)

        if (Settings.showDot) root.addView(roleDot)
        root.addView(label)
        addView(root)
    }

    fun setPermName(name: String) {
        label.text = name
    }

    fun setDotColor(color: Int) {
        val clr = if (color == 0) ColorCompat.getThemedColor(ctx, R.b.colorInteractiveNormal) else Color.parseColor("#%06x".format(color))
        val dotIcon = ContextCompat.getDrawable(ctx, R.e.drawable_circle_white_12dp)!!.mutate()
        dotIcon.setTint(clr)
        roleDot.setImageDrawable(dotIcon)
    }

    companion object {

        fun detailsButton(ctx: Context) = PermChip(ctx).apply {
            val shieldIcon = ContextCompat.getDrawable(ctx, R.e.ic_shieldstar_24dp)!!.mutate()
            shieldIcon.setTint(ColorCompat.getThemedColor(ctx, R.b.colorInteractiveNormal))

            label.visibility = GONE
            roleDot.apply {
                setImageDrawable(shieldIcon)
                (layoutParams as LinearLayout.LayoutParams).apply {
                    height = 16.dp
                    width = 16.dp
                    marginEnd = 0
                }
            }

            setOnClickListener {

            }
        }

    }

}